File: HW8/131044082/BigramDyn.java
import java.util.*;
import java.io.*;
 

public class BigramDyn<E> implements Bigram<E>
{
	private int dataType;
	private E[] allOfValuesInFileDyn;
	private int capacity=0;
	private int used=0;
	private Object value;
	
	//constructors
	public BigramDyn()
	{
		dataType = 2; // dataType hiç gelmez ise string ile çalışır.
	}

	public BigramDyn(int dType) throws Exception
	{
		dataType = dType;
		if(dataType != 1 && dataType != 2 && dataType != 3)
   			throw new Exception("Invalid dataType !!! ");  
		else
		{
			setCapacity(5);
			setUsed(0);
			allOfValuesInFileDyn = (E[]) new Object[capacity];
		}
	}

	//setters
	public void setCapacity(int newCapacity)
	{
		this.capacity = newCapacity;
	}

	public void setUsed (int newUsed)
	{
		this.used = newUsed;
	}

	//getters
	public int getCapacity()
	{
		return capacity;
	}

	public int getUsed()
	{
		return used;
	}

	//gönderilen filename den verileri okuyup oluşabilecek bütün bigramları hesaplar
	public void readFile(String fileName)
	{
		File input = new File(fileName);
		Scanner inputFile = null;

		try
		{
			inputFile = new Scanner(input);
			while(inputFile.hasNextLine())
			{
				Scanner s2 = new Scanner(inputFile.nextLine());
				
				while(s2.hasNext())
				{
					value = s2.next();
					allOfValuesInFileDyn[getUsed()] = (E) value;
					setUsed( getUsed() + 1 );					
					if(getUsed() >= getCapacity())
						controlOfCapacity();
				}
			}
		}
		catch(Exception e)
		{
			System.err.println(e);
		}
	}
	
	//bütün bigramların sayısını return eder
	public int numGrams()
	{
		return( getUsed() - 1 );
	}

	//parametre olarak aldığı bigram ikilisinden kaç tane olduğunu return eder
	public int numOfGrams(E anyType1, E anyType2)
	{
		int count = 0;

		for (int i = 0; i < getUsed() - 1; ++i)
		{
			if((String.valueOf(allOfValuesInFileDyn[i])).equals((String.valueOf(anyType1)))	&& (String.valueOf(allOfValuesInFileDyn[i+1])).equals((String.valueOf(anyType2))))	
				count++;
		}

		return count;
	}

	//arrayin içini azalan sırayla ekrana basar
	public String toString()
	{
		int[] intArray = new int[getUsed()];

		for (int i=0; i < getUsed(); ++i)
			intArray[i] = 0;


		int numberOfGram;
		for (int i = 0, j=1; i < getUsed(); ++i, ++j)
		{
			numberOfGram = numOfGrams(allOfValuesInFileDyn[i], allOfValuesInFileDyn[j]);
			intArray[i] = numberOfGram;
		}

		int temp;
		for (int i = 0; i < getUsed(); ++i)
		{
			for (int k = 0; k < getUsed(); ++k)
			{
				if(intArray[i] > intArray[k])
				{
					temp = intArray[k];
					intArray[k] = intArray[i];
					intArray[i] = temp;
				}
			}
		}

		System.out.println("----------------------------");

		for (int i = 0; i < getUsed(); i++)
		{
			for (int k = 0; k < getUsed()-1; k++)
			{
				if((numOfGrams(allOfValuesInFileDyn[k], allOfValuesInFileDyn[k+1]) == intArray[i]) && 
					(intArray[i] != intArray[i+1]))
					System.out.println("[" + allOfValuesInFileDyn[k] + " - " + allOfValuesInFileDyn[k+1] + "]" + " -----> " + numOfGrams(allOfValuesInFileDyn[k], allOfValuesInFileDyn[k+1]));
			}
		}
		
		System.out.println("----------------------------");

		return " ";
	}

	//eğer capacity dolduysa yeni capacity ile aldığı yeri büyütür
	public void controlOfCapacity()
	{
		E[] temp;
		temp = (E[]) new Object[capacity];

		for(int i=0; i < getCapacity(); ++i)
			temp[i] = allOfValuesInFileDyn[i];

		capacity += capacity;

		allOfValuesInFileDyn = (E[]) new Object[getCapacity()];

		for (int i=0; i < getUsed(); i++) 
			allOfValuesInFileDyn[i] = temp[i];	
	}
}

File: HW7/131044082/BigramDyn.cpp
/*				AHMET MUZAFFER DÜLGER
					  131044082
			DATE LAST MODIFIED: 29.12.2016
*/

#include "BigramDyn.h"

using namespace std;

template<class T>
BigramDyn<T>::BigramDyn(int dType) throw(exception)
{
	dataType = dType;
	if(dataType != 1 && dataType != 2 && dataType != 3)
		throw invalid_argument("Invalid data type!!!");

	setCapacity(100);
	setUsed(0);
	allOfValuesInFileDyn = new T[capacity]; 
}

template<class T>
BigramDyn<T>::BigramDyn(const BigramDyn& objectDyn) throw(exception)
{
	setCapacity(objectDyn.getCapacity());
	setUsed(objectDyn.getUsed());
	dataType = objectDyn.dataType;
	if(dataType != 1 && dataType != 2 && dataType != 3)
		throw invalid_argument("Invalid data type!!!");

	allOfValuesInFileDyn = new T[capacity];

	for (int i = 0; i < objectDyn.getUsed(); ++i)
		allOfValuesInFileDyn[i] = objectDyn.allOfValuesInFileDyn[i];
}

template<class T>
BigramDyn<T>::~BigramDyn()
{
	delete [] allOfValuesInFileDyn;
	allOfValuesInFileDyn = nullptr;
}

template<class T>
int BigramDyn<T>::getUsed()const
{
	return used;
}

template<class T>
int BigramDyn<T>::getCapacity()const
{
	return capacity;
}

template<class T>
void BigramDyn<T>::setUsed(const int otherUsed)
{
	used = otherUsed;
}

template<class T>
void BigramDyn<T>::setCapacity(const int otherCapacity)
{
	capacity = otherCapacity;
}

template<class T>
void BigramDyn<T>::readFile(const string& fileName) throw(exception)
{
	fstream inputFile;

	inputFile.open(fileName);

	if (inputFile.fail())
		throw invalid_argument("Cannot open file!!");
	else
	{	inputFile >> value;
		if (inputFile.eof())
        	throw invalid_argument("This file is empty!!!");
       	else if(inputFile.fail())
	 		throw invalid_argument("Invalid argument!!!");
		else
		{
			inputFile.close();
			inputFile.open(fileName);
			while(!inputFile.eof())
			{
				inputFile >> value;
				if(inputFile.fail())
	 				throw invalid_argument("Invalid argument!!!");
				if(getUsed() > getCapacity())
					controlOfCapacity();
				allOfValuesInFileDyn[getUsed()] = value;
				setUsed(getUsed()+1);
			}
		}
	}
}

template<class T>
int BigramDyn<T>::numGrams()const
{
	return (getUsed() - 1);
}

template<class T>
int BigramDyn<T>::numOfGrams(T anyType1, T anyType2)const
{
	int count = 0;
	for (int i = 0; i < getUsed(); ++i)
	{
		if(allOfValuesInFileDyn[i] == anyType1 && allOfValuesInFileDyn[i+1] == anyType2)
			count++;
	}
	return count;
}

template<class T>
pair<T,T> BigramDyn<T>::maxGrams()const
{
	int count=0, tempInMaxGramDyn=0;
	pair<T,T> returnValue[getUsed() - 1];

	for(int i=0; i<getUsed(); i++)
	{
		for(int j=0; j<getUsed(); j++)
		{
			if(allOfValuesInFileDyn[i] == allOfValuesInFileDyn[j] && 
				allOfValuesInFileDyn[i+1] == allOfValuesInFileDyn[j+1])
					tempInMaxGramDyn++;
			if(tempInMaxGramDyn > count)
			{
				count = tempInMaxGramDyn;
				returnValue[1] = make_pair(allOfValuesInFileDyn[i], allOfValuesInFileDyn[i+1]);				
			}
		}
		tempInMaxGramDyn = 0;
	}

	return (make_pair(returnValue[1].first, returnValue[1].second));
}

template<class T>
void BigramDyn<T>::printMyObject(ostream& outputStream)
{
	pair<T,T> tempPair[getUsed()];
	int intArray[getUsed()-1];

	for (int i = 0; i < getUsed(); ++i)
		intArray[i] = 0;

	int i=0, j=1;
	for (; i < getUsed(), j < getUsed(); i++, j++)
	{
		tempPair[i] = make_pair(allOfValuesInFileDyn[i], allOfValuesInFileDyn[j]);
	}

	for (int i = 0; i < getUsed(); ++i)
	{
		int numberOfGram;
		numberOfGram = numOfGrams(tempPair[i].first, tempPair[i].second);
		intArray[i] = numberOfGram;
	}

	int temp;
	for (int i = 0; i < getUsed(); ++i)
	{
		for (int k = 0; k < getUsed(); ++k)
		{
			if(intArray[i] > intArray[k])
			{
				temp = intArray[k];
				intArray[k] = intArray[i];
				intArray[i] = temp;
			}
		}
	}

	outputStream << "------------------" << endl;
	outputStream << "operator << " << endl << endl;
	for (int i = 0; i < getUsed(); i++)
	{
		for (int k = 0; k < getUsed(); k++)
		{
			if((numOfGrams(allOfValuesInFileDyn[k], allOfValuesInFileDyn[k+1]) == intArray[i]) && (intArray[i] != intArray[i+1]) && (k < getUsed()-1))
				outputStream << "[" << allOfValuesInFileDyn[k] << " - " << allOfValuesInFileDyn[k+1] << "]" << 
								" -----> " << numOfGrams(allOfValuesInFileDyn[k], allOfValuesInFileDyn[k+1]) << endl;
		}
	}
	outputStream << "------------------";
}

template<class T>
void BigramDyn<T>::controlOfCapacity()
{
	T *temp;
	temp = new T[getCapacity()];
	for (int i = 0; i < getCapacity(); ++i)
		temp[i] = allOfValuesInFileDyn[i];

	delete [] allOfValuesInFileDyn;
	capacity += capacity;
	allOfValuesInFileDyn = new T[getCapacity()];

	for (int i = 0; i < getUsed(); ++i)
		allOfValuesInFileDyn[i] = temp[i];

	delete [] temp;
}